internal/services/servicebus/migration/subscription.go

```diff
@@ -0,0 +1,132 @@
+package migration
+
+import (
+	"context"
+	"log"
+
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/resourcegroups"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/features"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/servicebus/parse"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+)
+
+var _ pluginsdk.StateUpgrade = ServiceBusSubscriptionV0ToV1{}
+
+type ServiceBusSubscriptionV0ToV1 struct{}
+
+func (ServiceBusSubscriptionV0ToV1) Schema() map[string]*pluginsdk.Schema {
+	s := map[string]*pluginsdk.Schema{
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ForceNew: true,
+		},
+
+		"auto_delete_on_idle": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+
+		"default_message_ttl": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+
+		"lock_duration": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+
+		"dead_lettering_on_message_expiration": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+		},
+
+		"dead_lettering_on_filter_evaluation_error": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+
+		"enable_batched_operations": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+		},
+
+		"max_delivery_count": {
+			Type:     pluginsdk.TypeInt,
+			Required: true,
+		},
+
+		"requires_session": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			ForceNew: true,
+		},
+
+		"forward_to": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+
+		"forward_dead_lettered_messages_to": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+
+		"status": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+	}
+
+	if !features.ThreePointOhBeta() {
+		s["topic_name"] = &pluginsdk.Schema{
+			Type:          pluginsdk.TypeString,
+			Optional:      true,
+			Computed:      true,
+			ForceNew:      true,
+			Deprecated:    `Deprecated in favor of "topic_id"`,
+			ConflictsWith: []string{"topic_id"},
+		}
+
+		s["namespace_name"] = &pluginsdk.Schema{
+			Type:          pluginsdk.TypeString,
+			Optional:      true,
+			Computed:      true,
+			ForceNew:      true,
+			Deprecated:    `Deprecated in favor of "topic_id"`,
+			ConflictsWith: []string{"topic_id"},
+		}
+
+		s["resource_group_name"] = &pluginsdk.Schema{
+			Type:          pluginsdk.TypeString,
+			Optional:      true,
+			Computed:      true,
+			ForceNew:      true,
+			ValidateFunc:  resourcegroups.ValidateName,
+			Deprecated:    `Deprecated in favor of "topic_id"`,
+			ConflictsWith: []string{"topic_id"},
+		}
+	}
+	return s
+}
+
+func (ServiceBusSubscriptionV0ToV1) UpgradeFunc() pluginsdk.StateUpgraderFunc {
+	return func(ctx context.Context, rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+		log.Printf("[DEBUG] Updating 'id' if resource identifier is not in camel-case")
```

Comment:
This is not needed as some users might be using the state generated since v3.0, which is guaranteed to be camel case.